devito/tools/utils.py

```diff
@@ -147,22 +149,51 @@ def single_or(l):
 
 
 def filter_ordered(elements, key=None):
-    """Filter elements in a list while preserving order.
+    """
+    Filter elements in a list while preserving order.
 
     Parameters
     ----------
     key : callable, optional
         Conversion key used during equality comparison.
     """
+    if isinstance(elements, types.GeneratorType):
+        elements = list(elements)
     seen = set()
     if key is None:
-        key = lambda x: x
+        try:
+            unordered, inds = np.unique(elements, return_index=True)
+            return unordered[np.argsort(inds)].tolist()
+        except:
+            return sorted(list(set(elements)), key=elements.index)
     return [e for e in elements if not (key(e) in seen or seen.add(key(e)))]
 
 
+def filter_ordered_index(elements, key=None):
```

Comment:
I think this is too highly specialised to exist and be considered a "tool".

Since in the only place where it's currently used we have `key=None`, can you not just

* drop this function
* use a one liner [here](https://github.com/devitocodes/devito/pull/1107/files#diff-5d662e9e9c311233ce7b83c1d18d291aR168)
```
np.unique(elements, return_index=True)[1].sort().tolist()
```

or if you really want it to be a tool, I'd at least call it
```
filter_indices
```
to draw a neat difference w.r.t. `filter_ordered`

Reply:
>to draw a neat difference w.r.t. filter_ordered

Sure. I prefer to keep it as a tool though.

>np.unique(elements, return_index=True)[1].sort().tolist()

`sort` and `np.argsort` are not the same from what I remember.

Reply:
the one liner is just syntactic sugar... I think I put on a single line exactly what you had split over multiple lines? 

I'll check again

Reply:
Nevermind, was thinking about something else, moved there, just needs two lines becasue `sort` returns nothing